### PR TITLE
Fix hangs and slow loading file bugs from AFL++ fuzzing.

### PR DESCRIPTION
**Pull request by AliceLR**
Fixes most (but not all) of the problems found in #294.

Fixed bugs affecting multiple formats:
* `scan_module`: tempo changes and other effects would prevent breaking the scan when >512 rows are played in a single pattern, causing an infinite loop.
* `scan_module`: modules with no valid orders could never break out of the scan loop.
* `scan_module`: infinite pattern loop detection (see `storlek_11.it`) only worked due to the scan count for the current row incidentally being the correct value when the >512 rows check breaks the scan. Other modules could still hang in situations like this, so now there's an explicit check to see if the current row's scan count has overflowed.
* `libxmp_load_sample`: when attempting to load a sample from file, if the file is currently positioned at the EOF, loading is skipped. This isn't a foolproof fix but it reduces the slowness and RAM consumption of a few inputs by quite a bit. (Samples that would run over the EOF but aren't at it could also be trimmed to help with this.)

Fixed bugs in loaders:
* `amf_load.c`: missing EOF checks in various places.
* `dbm_load.c`: missing EOF checks in `get_patt`.
* `fnk_load.c`: added a sanity check to make sure reported sample lengths aren't larger than the reported module length.
* `gal4_load.c`: missing EOF check in `get_patt`, usage of `printf` instead of `D_`.
* `gal5_load.c`: missing EOF check in `get_patt`, usage of `printf` instead of `D_`.
* `gdm_load.c`: missing EOF checks in the pattern scan and pattern load loops.
* `it_load.c`: missing EOF checks in pattern scan and pattern load loops.
* `masi_load.c`: scan of subchunks in `get_song_2` could get caught in an infinite seeking loop due to very large chunk sizes being interpreted as negative sizes and not being filtered out.
* `mod_load.c`: missing `hio_read` return value checks could result in uninitialized reads.
* `stx_load.c`: missing EOF check in the pattern load loop.

Fixed bugs in depackers:
* `gunzip.c`: missing EOF/error checks in the name and comment skipping loops could cause hangs.
* `inflate.c`: multiple missing EOF/error checks in `decompress` could cause hangs.
* `unzip.c`: missing EOF/error checks in the store handling could cause hangs.

### CODE REVIEW

**Review by sezero — APPROVED** [review]

**Comment by sezero** [comment]
With this I'm getting a warning from scan.c which hasn't been there before:
```
src/scan.c: In function ‘scan_module’:
src/scan.c:57: warning: ‘row’ may be used uninitialized in this function
```
This is with my old gcc4.4.7.

Other than that, looks good.

**Comment by AliceLR** [comment]
> With this I'm getting a warning from scan.c which hasn't been there before:
> 
> ```
> src/scan.c: In function ‘scan_module’:
> src/scan.c:57: warning: ‘row’ may be used uninitialized in this function
> ```
> 
> This is with my old gcc4.4.7.
> 
> Other than that, looks good.

I was able to reproduce this with GCC 4.9.5 using -O1 and it's because I should have used `break` instead of `goto end_module` in the invalid orders loop breaking check. This should be fixed now.

File: src/scan.c [hunk 1]
```diff
@@ -116,11 +117,19 @@ static int scan_module(struct context_data *ctx, int ep, int chain)
...
-    gvol_memory = break_row = row_count = frame_count = 0;
+    gvol_memory = break_row = row_count = row_count_total = frame_count = 0;
+    orders_since_last_valid = any_valid = 0;
     start_time = time = 0.0;
     inside_loop = 0;
 
     while (42) {
+	/* Sanity check to prevent getting stuck due to broken patterns. */
+	if (orders_since_last_valid > 512) {
+	    D_(D_CRIT "orders_since_last_valid = %d @ ord %d; ending scan", orders_since_last_valid, ord);
```
Comment (sezero) [inline src/scan.c:128]: Should this imply a failure?  I.e. should we set `any_valid` to 0 here?

Comment (AliceLR) [reply src/scan.c:128]: I don't think so—some valid tracks (particularly ones containing a lot of loops) are capable of reaching this. This OctaMED module should loop for something like an hour(!) and encounters this check, but it is a valid module.

https://modland.com/pub/modules/OctaMED%20MMD1/Apollon/pilot%20i%20tyskland.mmd1

Comment (AliceLR) [reply src/scan.c:128]: Sorry, I deleted a comment I made thinking this was the other >512 check. I think it theoretically might be possible to cause this check to occur and still have a partial valid module. I'm not entirely sure.

Comment (sezero) [reply src/scan.c:128]: Will you investigate more, or should we merge?

Comment (AliceLR) [reply src/scan.c:128]: I tried yesterday to create a test case that could cause that state to occur but failed to, so I think I'll probably have to step through the code. You made a valid point and what I suspect could occur is that if this check is flagged but allows the scan to succeed (`any_valid`=1), the player scan could later get stuck in an infinite loop (a problem I encountered before adding `any_valid` at all). So the solution may be to remove `any_valid` and just unconditionally return -1 here.

Comment (AliceLR) [reply src/scan.c:128]: Okay, after some experimentation, I think that the current implementation should be fine—as long as the sequence has a valid entry point it won't get stuck in an infinite loop in `next_order` and `any_valid` makes sure that the entry point (eventually) reaches something valid.

Comment (sezero) [reply src/scan.c:128]: OK then.

**Pull request merged**